File: mobile/welcome-screen/shared/ios/TutorialShell.swift
//  TutorialShell.swift
//  ShowUp · the shared chrome for every tutorial screen (SHOWUP-135)
//
//  Screen 2 ("Meet in real life") is the reference implementation of this chrome; screens 3-6 reuse
//  it unchanged and pass content only. Nothing screen-specific belongs in this file.
//
//  The layout rule that matters: this is ONE top-anchored column with a SINGLE flexible region
//  between the illustration and the nav row. That region is the only thing that absorbs height
//  differences between screens and between devices. No Y position is ever hard-coded — offsets that
//  look right at 844 tall break on every other frame.
//
//  The illustration is the only flexible element. On short frames it shrinks; the type, the nav row
//  and the safe-area margins never do, and the screen never scrolls.

import SwiftUI
import UIKit

// MARK: - ② Step progress — 5 segments, height 5, gap 6, full content width

struct StepProgress: View {
    let steps: Int
    let current: Int
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    var body: some View {
        HStack(spacing: 6) {
            ForEach(0..<steps, id: \.self) { i in
                Capsule()
                    .fill(i < current ? Color.liqPurple : Color.liqTrack)
                    .frame(height: 5)
            }
        }
        // Advancing a card should read as progress being made, not as the bar being redrawn.
        .animation(reduceMotion ? nil : .easeInOut(duration: 0.32), value: current)
        // Five anonymous capsules carry no text: without this VoiceOver announces nothing here.
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text("Step \(current) of \(steps)"))
    }
}

// MARK: - ③ Eyebrow pill — Manrope 700 / 11 / uppercase, tracking .08, padding 5 / 10
//
// Placed by the caller inside an HStack with a trailing Spacer, so it hugs its text rather than
// stretching to the content width. The reference render shows the stretched version; the design
// system component is hug-width.

struct EyebrowPill: View {
    let text: String

    var body: some View {
        HStack(spacing: 6) {
            Circle().fill(Color.liqOrange).frame(width: 5, height: 5)
            Text(text.uppercased())
                .font(F.manrope(11, .bold))
                .tracking(0.08 * 11)
                .foregroundColor(.liqPurple)
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 5)
        .background(Capsule().fill(Color.liqEyebrowBg))
    }
}

// MARK: - ⑨ Next — label plus a 56pt circular arrow, gap 14

/// Which circle the nav row's forward action wears.
///
/// A variant on the shared button, not a forked nav row — screen 6 is the only terminal screen, and
/// forking would mean its progress bar and Back slot drift from the other five.
enum NextVariant { case orange, sunset }

/// The spec's arrow: a 2pt stroke with round caps, not a filled glyph.
///
/// `Image(systemName: "arrow.right")` renders at the system weight with a different silhouette;
/// at 56pt against a saturated circle that difference shows. Drawing it also keeps iOS and Android
/// pixel-identical, which an SF Symbol never can.
struct ArrowRight: Shape {
    func path(in r: CGRect) -> Path {
        var p = Path()
        let s = r.width
        let midY = r.midY
        p.move(to: CGPoint(x: s * 0.10, y: midY))
        p.addLine(to: CGPoint(x: s * 0.84, y: midY))
        p.move(to: CGPoint(x: s * 0.56, y: midY - s * 0.25))
        p.addLine(to: CGPoint(x: s * 0.86, y: midY))
        p.addLine(to: CGPoint(x: s * 0.56, y: midY + s * 0.25))
        return p
    }
}

/// Publishes the button's pressed state down the environment so only the circle scales and the
/// label stays put — matching Android, where the press transform sits on the circle alone.
private struct IsPressedKey: EnvironmentKey { static let defaultValue = false }
extension EnvironmentValues {
    var nextIsPressed: Bool {
        get { self[IsPressedKey.self] }
        set { self[IsPressedKey.self] = newValue }
    }
}
private struct PressReporter: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label.environment(\.nextIsPressed, configuration.isPressed)
    }
}

private struct NextCircle: View {
    let variant: NextVariant
    @Environment(\.nextIsPressed) private var isPressed
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private var fill: AnyShapeStyle {
        switch variant {
        case .orange:
            return AnyShapeStyle(Color.liqOrange)
        case .sunset:
            // 135 degrees, midpoint at 38% - deliberately not an even three-stop ramp
            return AnyShapeStyle(LinearGradient(
                stops: [
                    .init(color: .liqOrange, location: 0.00),
                    .init(color: Color(hex: 0xD05976), location: 0.38),
                    .init(color: .liqPurple, location: 1.00),
                ],
                startPoint: .topLeading, endPoint: .bottomTrailing))
        }
    }
    private var glow: Color { variant == .sunset ? .liqPurple : .liqOrange }
    /// Card 05's arrow is 22; cards 01-04 use 20.
    private var arrowSize: CGFloat { variant == .sunset ? 22 : 20 }

    var body: some View {
        ZStack {
            Circle().fill(fill).frame(width: 56, height: 56)
            ArrowRight()
                .stroke(Color.white,
                        style: StrokeStyle(lineWidth: 2, lineCap: .round, lineJoin: .round))
                .frame(width: arrowSize, height: arrowSize)
        }
        .shadow(color: glow.opacity(0.5), radius: 16, y: 10)
        // 0.94 reads as a press without the circle appearing to shrink away from the finger.
        .scaleEffect(isPressed && !reduceMotion ? 0.94 : 1)
        .animation(.spring(response: 0.24, dampingFraction: 0.55), value: isPressed)
    }
}

struct NextButton: View {
    let label: String
    var variant: NextVariant = .orange
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 14) {
                Text(label)
                    .font(F.manrope(17, .bold))
                    .foregroundColor(.liqFg)
                NextCircle(variant: variant)
            }
        }
        .buttonStyle(PressReporter())
        .accessibilityLabel(Text(label))
        .accessibilityAddTraits(.isButton)
    }
}

// MARK: - The shell

/// Everything here is identical across tutorial screens 2-6.
///
/// - Parameters:
///   - step: which segment of the progress bar is filled (1-based)
///   - totalSteps: how many segments — 5 for the tour
///   - showBack: false on the first screen of the tour. The slot stays, reserving layout width,
///     but is invisible and non-interactive.
///   - headline: passed in so each screen styles its own italic run
///   - content: ⑤⑥⑦ the per-screen body — rule rows, paragraphs
///   - art: ⑧ the illustration block, the only element allowed to shrink
struct TutorialShell<Headline: View, Content: View, Art: View>: View {
    let step: Int
    let totalSteps: Int
    let eyebrow: String
    let nextLabel: String
    var nextVariant: NextVariant = .orange
    var showBack: Bool = true
    /// Width of the headline underline accent. It underlines the italic phrase, which sits in a
    /// different place in every headline, so the width is per-card rather than derived. 0 draws
    /// nothing.
    var underlineWidth: CGFloat = 0
    var onNext: () -> Void = {}
    var onBack: () -> Void = {}
    @ViewBuilder let headline: () -> Headline
    @ViewBuilder let content: () -> Content
    @ViewBuilder let art: () -> Art

    var body: some View {
        ZStack {
            Color.liqCream.ignoresSafeArea()

            VStack(alignment: .leading, spacing: 0) {
                StepProgress(steps: totalSteps, current: step)
                    .padding(.top, 8)                      // ① pad-top 8 below the safe-area inset

                Spacer().frame(height: 24)                 // progress -> eyebrow

                HStack(spacing: 0) {                       // hug-width, not a full-width band
                    EyebrowPill(text: eyebrow)
                    Spacer(minLength: 0)
                }

                Spacer().frame(height: 14)                 // eyebrow -> headline

                // ④ headline + the underline accent every card's AC calls for. It was on the
                // Welcome screen only.
                //
                // An overlay, so it cannot affect the headline's measured height - the 14 / 28
                // margins around it are untouched.
                headline()
                    .overlay(alignment: .bottomLeading) {
                        if underlineWidth > 0 {
                            LinearGradient(
                                stops: [
                                    .init(color: .clear, location: 0.00),
                                    .init(color: Color.liqOrange.opacity(0.50), location: 0.28),
                                    .init(color: Color(hex: 0xE0567A).opacity(0.40), location: 0.64),
                                    .init(color: .clear, location: 1.00),
                                ],
                                startPoint: .leading, endPoint: .trailing
                            )
                            .frame(width: underlineWidth, height: 10)
                            .clipShape(Capsule())
                            .offset(y: 4)
                            .allowsHitTesting(false)
                            .accessibilityHidden(true)
                        }
                    }

                Spacer().frame(height: 28)                 // headline -> content

                content()

                Spacer().frame(height: 24)                 // content -> illustration

                // The single flexible region. Everything above is fixed; the art takes what is
                // left, capped at its natural height, and the surplus below it pushes the nav row
                // to the bottom.
                art()
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)

                // ⑨ nav row — bottom-anchored, 24 above the content floor
                HStack {
                    // 44pt is Apple's minimum touch target; the label alone is ~17pt tall. The
                    // frame carries the target and the text sits at its leading edge - identical
                    // pixels, a control that passes an accessibility audit. While the slot is
                    // reserved but invisible (card 01) it is hidden from VoiceOver too, so nothing
                    // announces a "Back" that cannot be pressed.
                    Button(action: { if showBack { onBack() } }) {
                        Text("Back")
                            .font(F.manrope(14, .semibold))
                            .foregroundColor(showBack ? .liqSubtle : .clear)
                            .frame(minWidth: 44, minHeight: 44, alignment: .leading)
                            .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .disabled(!showBack)
                    .accessibilityHidden(!showBack)

                    Spacer()
                    NextButton(label: nextLabel, variant: nextVariant, action: onNext)
                }
                .padding(.bottom, 24)
            }
            .padding(.horizontal, 24)                      // gutter 24
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
        }
    }
}

// MARK: - ⑤ Rule row — shared by screens 2-5: dot · rule · faint dash · purple consequence

struct RuleRow: View {
    let rule: String
    let consequence: String
    var wraps: Bool = true

    var body: some View {
        HStack(alignment: .top, spacing: 9) {
            Circle()
                .fill(Color.liqOrange)
                .frame(width: 6, height: 6)
                .padding(.top, 6)              // sits on the first line's optical centre
            Text(TypeMetrics.attributed(
                colouredRuns: [
                    (rule, TypeMetrics.uiFont(PS.manropeSemi, 13,
                                              fallback: .systemFont(ofSize: 13, weight: .semibold)),
                     UIColor(Color.liqFg)),
                    (" — ", TypeMetrics.uiFont(PS.manropeMedium, 13,
                                               fallback: .systemFont(ofSize: 13, weight: .medium)),
                     UIColor(Color.liqFaint)),
                    (consequence, TypeMetrics.uiFont(PS.manropeSemi, 13,
                                                     fallback: .systemFont(ofSize: 13, weight: .semibold)),
                     UIColor(Color.liqPurple)),
                ],
                size: 13, multiple: 1.4, trackingEm: -0.01
            ))
            .lineLimit(wraps ? nil : 1)          // screen 2 is specced nowrap; 3-5 wrap
            .fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
        }
    }
}

// MARK: - ⑧ Placeholder for the artwork design will supply
//
// It holds the exact block the real asset gets — 248 x 210, centred, shrinking with the frame — so
// dropping the image in later changes nothing about the layout. Replace the placeholder with
// Image(...).resizable().scaledToFit() and the surrounding code is unchanged.

struct IllustrationPlaceholder: View {
    /// Screen 6 draws at 0.62 — its text block is the tallest of the set.
    var scale: CGFloat = 1

    var body: some View {
        ZStack {
            // radial glow — kept so the "no banding" criterion stays testable
            RadialGradient(
                gradient: Gradient(stops: [
                    .init(color: .liqOrange.opacity(0.16), location: 0.00),
                    .init(color: .liqPurple.opacity(0.10), location: 0.48),
                    .init(color: .liqPurple.opacity(0.00), location: 0.70),
                ]),
                center: .center, startRadius: 0, endRadius: 120
            )
            .blur(radius: 6)

            GeometryReader { geo in
                let side = min(geo.size.height, 230 * scale)
                RoundedRectangle(cornerRadius: 18)
                    .strokeBorder(
                        Color.liqPurple.opacity(0.38),
                        style: StrokeStyle(lineWidth: 1.5, dash: [7, 6])
                    )
                    .overlay(
                        VStack(spacing: 3) {
                            Image(systemName: "photo")
                                .font(.system(size: 24, weight: .light))
                                .foregroundColor(.liqPurple.opacity(0.5))
                            Text("ILLUSTRATION")
                                .font(F.manrope(11, .bold))
                                .tracking(0.07 * 11)
                                .foregroundColor(.liqPurple.opacity(0.75))
                            Text("248 × 210")
                                .font(F.manrope(10.5, .semibold))
                                .foregroundColor(.liqSubtle)
                        }
                    )
                    .frame(width: side * (248.0 / 210.0), height: side)
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
            }
        }
        .frame(maxHeight: 230 * scale)
        // Decorative: it carries no information the copy does not already state, so VoiceOver
        // skips it rather than being given a label it would have to read past.
        .accessibilityHidden(true)
    }
}


// MARK: - Screen 6 only: a plain statement row, single colour
//
// Deliberately NOT the two-tone rule/consequence pattern — that sells a benefit, this states
// policy. 14.5 / 1.42, 7pt dot at offset 7, gap 12 to the text.

struct StatementRow: View {
    let text: String

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Circle().fill(Color.liqOrange).frame(width: 7, height: 7).padding(.top, 7)
            Text(TypeMetrics.attributed(
                runs: [(text, TypeMetrics.uiFont(PS.manropeSemi, 14.5,
                        fallback: .systemFont(ofSize: 14.5, weight: .semibold)))],
                size: 14.5, multiple: 1.42, color: UIColor(Color.liqFg)))
                .fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
        }
    }
}

File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/tutorial/TutorialShell.kt
/*
 * TutorialShell.kt
 * ShowUp · the shared chrome for every tutorial screen (SHOWUP-135)
 *
 * Screen 2 ("Meet in real life") is the reference implementation of this chrome; screens 3-6 reuse
 * it unchanged and pass content only. Nothing screen-specific belongs in this file.
 *
 * The layout rule that matters: this is ONE top-anchored column with a SINGLE Spacer(weight = 1f)
 * between the illustration and the nav row. That spacer is the only thing that absorbs height
 * differences between screens and between devices. No Y position is ever hard-coded — offsets that
 * look right at 844 tall break on every other frame.
 *
 * The illustration is the only flexible element. On short frames it shrinks; the type, the nav row
 * and the safe-area margins never do, and the screen never scrolls.
 */
package com.showup.tutorial

import com.showup.designsystem.IconSizes
import com.showup.designsystem.Spacing

import androidx.compose.foundation.Canvas
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.progressBarRangeInfo
import androidx.compose.ui.semantics.ProgressBarRangeInfo
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.unit.Dp
import android.provider.Settings
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import com.showup.designsystem.Cream
import com.showup.welcome.ctaGlow
import com.showup.designsystem.EyebrowBg
import com.showup.designsystem.Faint
import com.showup.designsystem.Fg
import com.showup.designsystem.Manrope
import com.showup.designsystem.Orange
import com.showup.designsystem.Purple
import com.showup.designsystem.Subtle
import com.showup.designsystem.Track

/**
 * ② Step progress — 5 segments, height 5, gap 6, full content width.
 *
 * The segment colour animates rather than snapping, so advancing a card reads as progress being
 * made rather than as the bar being redrawn. It is one colour tween per segment, which costs
 * nothing and is skipped entirely when the device asks for no motion.
 *
 * Semantics: the bar is one node reporting "Step N of 5", not five anonymous boxes. Without this a
 * screen reader announces nothing at all here — the segments carry no text.
 */
@Composable
fun StepProgress(steps: Int, current: Int, modifier: Modifier = Modifier) {
    val motion = rememberMotion()
    Row(
        modifier
            .fillMaxWidth()
            .semantics(mergeDescendants = true) {
                progressBarRangeInfo =
                    ProgressBarRangeInfo(current.toFloat(), 0f..steps.toFloat(), steps)
                contentDescription = "Step " + current + " of " + steps
            },
        horizontalArrangement = Arrangement.spacedBy(Spacing.sm),
    ) {
        repeat(steps) { i ->
            val target = if (i < current) Purple else Track
            val segment by animateColorAsState(
                targetValue = target,
                animationSpec = tween(durationMillis = if (motion.enabled) 320 else 0),
                label = "segment",
            )
            Box(
                Modifier
                    .weight(1f)
                    .height(5.dp)
                    .clip(RoundedCornerShape(50))
                    .background(segment)
            )
        }
    }
}

/**
 * ③ Eyebrow pill — Manrope 700 / 11 / uppercase, tracking .08, padding 5/10.
 *
 * `align(Alignment.Start)` is load-bearing: in a Column the default stretches children to the full
 * content width, which is the full-width band the reference render shows. The design-system
 * component is a hug-width pill.
 *
 * CONFIRMED 2026-09-01, after this was queried a second time for looking too short next to the
 * render. It is short on purpose. The 04-thirty-minutes spec sheet says so in its own words:
 *
 *     "intended hug width (align-self flex-start) -- reference render still shows the
 *      full-width stretch, as on card 01"
 *
 * So the render disagreeing with us is the discrepancy the handoff already knows about, not a bug
 * on our side. Anyone comparing the two will notice it again; this note is here so the next person
 * spends a minute on it rather than an afternoon. Changing it means overruling the spec text, which
 * is the product side's call and is recorded in audit/CONFLICTS-2026-08-27.md as A10.
 */
@Composable
fun ColumnScope.EyebrowPill(text: String, modifier: Modifier = Modifier) {
    Row(
        modifier
            .align(Alignment.Start)
            .clip(RoundedCornerShape(50))
            .background(EyebrowBg)
            .padding(horizontal = Spacing.lg, vertical = 5.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.sm),
    ) {
        Box(Modifier.size(5.dp).background(Orange, CircleShape))
        Text(
            text.uppercase(),
            color = Purple, fontFamily = Manrope, fontWeight = FontWeight.Bold,
            fontSize = 11.sp, lineHeight = 13.sp, letterSpacing = 0.08.em,
        )
    }
}

/**
 * Which circle the nav row's forward action wears.
 *
 * A variant on the shared button, not a forked nav row — screen 6 is the only terminal screen, and
 * forking would mean its progress bar and Back slot drift from the other five.
 */
enum class NextVariant { Orange, Sunset }

/**
 * Whether this device wants motion.
 *
 * Android has no single "reduce motion" flag. The honest signal is the system animator duration
 * scale, which the OS sets to 0 when someone turns animations off — either in Accessibility >
 * Remove animations, or in Developer options. Respecting it keeps the flow usable for people who
 * get motion sick, and has the useful side effect of holding the screens still under UI tests.
 */
@Immutable
data class Motion(val enabled: Boolean)

@Composable
fun rememberMotion(): Motion {
    val context = LocalContext.current
    return remember(context) {
        val scale = runCatching {
            Settings.Global.getFloat(context.contentResolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f)
        }.getOrDefault(1f)
        Motion(enabled = scale > 0f)
    }
}

/**
 * The spec's arrow — a 2px stroke with round caps, not a filled glyph.
 *
 * Material's `Icons.AutoMirrored.Filled.ArrowForward` is a solid shape with a different silhouette;
 * at 56dp against a saturated circle the difference is plainly visible, so the arrow is drawn.
 */
@Composable
private fun ArrowRight(size: Dp) {
    Canvas(Modifier.size(size)) {
        val s = this.size.width
        val midY = this.size.height / 2f
        val w = 2.dp.toPx()
        drawLine(
            Color.White, Offset(s * 0.10f, midY), Offset(s * 0.84f, midY),
            strokeWidth = w, cap = StrokeCap.Round,
        )
        drawPath(
            Path().apply {
                moveTo(s * 0.56f, midY - s * 0.25f)
                lineTo(s * 0.86f, midY)
                lineTo(s * 0.56f, midY + s * 0.25f)
            },
            Color.White,
            style = Stroke(width = w, cap = StrokeCap.Round, join = StrokeJoin.Round),
        )
    }
}

/**
 * ⑨ Next — label plus a 56dp circular arrow, gap 14.
 *
 * Card 05 is the only terminal card. Its circle takes the sunset gradient and a violet shadow, and
 * its arrow is 22 rather than 20 — both variants on this one button, never a forked nav row.
 */
@Composable
fun NextButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: NextVariant = NextVariant.Orange,
) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val motion = rememberMotion()
    // 0.94 reads as a press without the circle appearing to shrink away from the finger.
    val scale by animateFloatAsState(
        targetValue = if (pressed && motion.enabled) 0.94f else 1f,
        animationSpec = spring(dampingRatio = 0.55f, stiffness = 900f),
        label = "nextPress",
    )
    val glow = if (variant == NextVariant.Sunset) Purple else Orange
    Row(
        modifier
            .clip(RoundedCornerShape(50))
            .clickable(
                interactionSource = interaction,
                indication = null,          // the scale IS the feedback; a ripple would fight it
                role = Role.Button,
                onClick = onClick,
            ),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Text(label, color = Fg, fontFamily = Manrope, fontWeight = FontWeight.Bold, fontSize = 17.sp)
        Box(
            Modifier
                .graphicsLayer { scaleX = scale; scaleY = scale }
                // --liq-shadow-cta on cards 01-04, --liq-shadow-violet on card 05.
                //
                // Drawn, not cast. Modifier.shadow uses Android's elevation system, whose light
                // sits at the top-centre of the WINDOW -- so the shadow's direction depends on
                // where the control happens to sit on screen, and this circle lives at the right
                // edge, which threw its glow down and to the LEFT. The token is a glow: no light
                // source, no direction, spread evenly and pushed straight down.
                .size(IconSizes.badge)
                .ctaGlow(glow)
                .clip(CircleShape)
                .then(
                    when (variant) {
                        NextVariant.Orange -> Modifier.background(Orange)
                        // 135 degrees, midpoint at 38% - deliberately not an even three-stop ramp
                        NextVariant.Sunset -> Modifier.background(
                            Brush.linearGradient(
                                0.00f to Orange,
                                0.38f to Color(0xFFD05976),
                                1.00f to Purple,
                            )
                        )
                    }
                ),
            contentAlignment = Alignment.Center,
        ) {
            ArrowRight(size = if (variant == NextVariant.Sunset) 22.dp else 20.dp)
        }
    }
}

/**
 * ⑤ A rule row: dot · rule · faint dash · purple consequence.
 *
 * Shared because four screens use it. Screen 2 sets `wraps = false` (spec: nowrap, single line);
 * screens 3-5 wrap to two lines, which is why the flag exists rather than a hard-coded rule.
 */
@Composable
fun RuleRow(rule: String, consequence: String, wraps: Boolean = true) {
    Row(verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(9.dp)) {
        // 6dp dot, offset 6 from the top so it sits on the first line's optical centre
        Box(Modifier.padding(top = Spacing.sm).size(6.dp).background(Orange, CircleShape))
        Text(
            buildAnnotatedString {
                append(rule)
                withStyle(SpanStyle(color = Faint, fontWeight = FontWeight.Medium)) { append(" — ") }
                withStyle(SpanStyle(color = Purple)) { append(consequence) }
            },
            color = Fg, fontFamily = Manrope, fontWeight = FontWeight.SemiBold,
            fontSize = 13.sp, lineHeight = 18.2.sp, letterSpacing = (-0.01).em,
            maxLines = if (wraps) Int.MAX_VALUE else 1,
            softWrap = wraps,
        )
    }
}

/**
 * Screen 6 only: a plain statement row, single colour.
 *
 * Deliberately NOT the two-tone rule/consequence pattern — that sells a benefit, this states
 * policy. 14.5 / 1.42, 7dp dot at offset 7, gap 12 to the text.
 */
@Composable
fun StatementRow(text: String) {
    Row(verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(Spacing.xl)) {
        Box(Modifier.padding(top = 7.dp).size(7.dp).background(Orange, CircleShape))
        Text(
            text, color = Fg, fontFamily = Manrope, fontWeight = FontWeight.SemiBold,
            fontSize = 14.5.sp, lineHeight = 20.6.sp,
        )
    }
}

/**
 * The screen shell. Everything here is identical across tutorial screens 2-6.
 *
 * @param step          which segment of the progress bar is filled (1-based)
 * @param totalSteps    how many segments — 5 for the tour
 * @param eyebrow       the pill label
 * @param showBack      false on the first screen of the tour: the slot stays, reserving layout
 *                      width, but is invisible and non-interactive
 * @param headline      the Lora headline, passed in so each screen can style its own italic run
 * @param content       ⑤⑥⑦ the per-screen body — rule rows, paragraphs
 * @param art           ⑧ the illustration block, 342 x 230, the only element allowed to shrink.
 *                      BoxScope, not ColumnScope: art is invoked inside the flexible Box below.
 *                      Compose marks every layout scope with @LayoutScopeMarker (a @DslMarker), so
 *                      opening a Box scope hides the enclosing ColumnScope — a ColumnScope receiver
 *                      here has nothing to resolve against and fails to compile.
 */
@Composable
fun TutorialShell(
    step: Int,
    totalSteps: Int,
    eyebrow: String,
    nextLabel: String,
    onNext: () -> Unit,
    nextVariant: NextVariant = NextVariant.Orange,
    modifier: Modifier = Modifier,
    showBack: Boolean = true,
    onBack: () -> Unit = {},
    headline: @Composable ColumnScope.() -> Unit,
    content: @Composable ColumnScope.() -> Unit,
    art: @Composable BoxScope.() -> Unit,
) {
    Box(modifier.fillMaxSize().background(Cream)) {
        Column(
            Modifier
                .fillMaxSize()
                .windowInsetsPadding(WindowInsets.safeDrawing)
                // ① content region: pad-top 8 below the safe-area inset, gutter 24, floor 0
                .padding(start = Spacing.screenGutter, end = Spacing.screenGutter, top = Spacing.md),
        ) {
            StepProgress(totalSteps, step)
            Spacer(Modifier.height(24.dp))          // progress -> eyebrow

            EyebrowPill(eyebrow)
            Spacer(Modifier.height(14.dp))          // eyebrow -> headline

            // ④ headline + the underline accent. Every card's AC asks for it; only the Welcome
            //    screen had one.
            //
            //    The bar is a child of a Box wrapping just the headline, bottom-aligned and nudged
            //    down 4. Modifier.offset places without re-measuring, so the accent cannot change
            //    the headline's measured height — the column's 14 / 28 margins are untouched.
            Column { headline() }
            Spacer(Modifier.height(28.dp))          // headline -> content

            content()

            Spacer(Modifier.height(24.dp))          // content -> illustration

            // The single flexible region. Compose does not shrink a fixed-height child the way CSS
            // flex does, so rather than "art at 230 + a spacer", the art's SLOT takes all remaining
            // space and the art measures min(available, 230) inside it, aligned to the top. Same
            // result as the spec's "art + flex:1 spacer": art at its natural size on tall frames,
            // shrinking on short ones, nav row pinned to the bottom either way.
            Box(
                Modifier.weight(1f).fillMaxWidth(),
                contentAlignment = Alignment.TopCenter,
            ) {
                art()
            }

            // ⑨ nav row — bottom-anchored, 24 above the content floor
            Row(
                Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // The label is ~20dp tall; 48dp is Android's minimum touch target, so the box
                // carries the target and the text sits at its leading edge — same pixels, a
                // tappable area that passes an accessibility scan. When the slot is reserved but
                // invisible (card 01) it is also cleared from the semantics tree, so a screen
                // reader does not announce a "Back" that cannot be pressed.
                Box(
                    Modifier
                        .defaultMinSize(minWidth = 48.dp, minHeight = 48.dp)
                        .then(
                            if (showBack) Modifier
                                .clip(RoundedCornerShape(50))
                                .clickable(role = Role.Button, onClick = onBack)
                            else Modifier.clearAndSetSemantics { }
                        ),
                    contentAlignment = Alignment.CenterStart,
                ) {
                    Text(
                        "Back",
                        color = if (showBack) Subtle else Color.Transparent,
                        fontFamily = Manrope, fontWeight = FontWeight.SemiBold, fontSize = 14.sp,
                    )
                }
                NextButton(nextLabel, onNext, variant = nextVariant)
            }
            Spacer(Modifier.height(24.dp))
        }
    }
}

/**
 * ⑧ Placeholder for the artwork design will supply.
 *
 * It holds the exact block the real asset gets — 248 x 210, centred, shrinking with the frame — so
 * dropping the image in later changes nothing about the layout. Replace the Canvas with an
 * Image(painterResource(...)) and the surrounding code is unchanged.
 *
 * `scale` exists for screen 6, which is drawn at 0.62 because its text block is the tallest of the
 * set and would not otherwise fit.
 */
@Composable
fun IllustrationPlaceholder(scale: Float = 1f) {
    // Decorative: it carries no information the copy does not already state, so it is cleared from
    // the semantics tree rather than given a label a screen reader would have to read past.
    Box(contentAlignment = Alignment.Center, modifier = Modifier.clearAndSetSemantics { }) {
        // Radial glow — kept so the "no banding" criterion stays testable.
        //
        // Three stops exactly as the reference: sunset .16 at the centre, violet .10 at 48%,
        // transparent by 70%. An earlier build carried a fourth stop at 26% that exists in no
        // source; it flattened the falloff and is the kind of drift that makes a gradient
        // "nearly right" forever.
        Canvas(Modifier.matchParentSize()) {
            drawCircle(
                brush = Brush.radialGradient(
                    0.00f to Orange.copy(alpha = 0.16f),
                    0.48f to Purple.copy(alpha = 0.10f),
                    0.70f to Color.Transparent,
                    1.00f to Color.Transparent,
                ),
                radius = size.minDimension * 0.62f,
            )
        }
        Box(
            Modifier.fillMaxHeight().heightIn(max = 230.dp * scale).aspectRatio(248f / 210f),
            contentAlignment = Alignment.Center,
        ) {
            Canvas(Modifier.matchParentSize()) {
                drawRoundRect(
                    color = Purple.copy(alpha = 0.38f),
                    cornerRadius = CornerRadius(18.dp.toPx()),
                    style = Stroke(
                        width = 1.5.dp.toPx(),
                        pathEffect = PathEffect.dashPathEffect(
                            floatArrayOf(7.dp.toPx(), 6.dp.toPx()), 0f,
                        ),
                    ),
                )
            }
            Text(
                "ILLUSTRATION\n248 × 210",
                color = Purple.copy(alpha = 0.75f), fontFamily = Manrope,
                fontWeight = FontWeight.Bold, fontSize = 11.sp, lineHeight = 16.sp,
                letterSpacing = 0.07.em,
            )
        }
    }
}
